### ISSUE TITLE
client repeats first "acted" message when service-process restarts

### ISSUE THREAD

**Issue by platdesign** [senecajs/seneca#318]
i have a strange issue. i have built a service (listening on a tcp port) and a client which connects to the service. i can add methods to the service and act them from the client. everything fine so far. but everytime i restart the service-process it receives the first "acted" method (since client-process started) from the client again and executes it. after restarting the client-process the service does not receive messages when it is restarted. but after the first act-call on the client, the service-process receives the first "acted" message again. any ideas?

**service.js**
```javascript
'use strict';

var seneca = require('seneca');

var service = seneca();

service.add({
	role: 'foo',
	cmd: 'bar'
}, function(args, cb) {
	console.log(args.name, Date.now(), args.meta$.id);
	cb(null, { test:123, name: args.name });
} )


service.listen({
	type: 'tcp',
	port: 12345,
	host: '0.0.0.0'
});

```

**client.js**
```javascript
'use strict';

var seneca = require('seneca');

var service = seneca();

service.client({
	type: 'tcp',
	port: 12345,
	host: '0.0.0.0',
	pin: 'role:foo'
});


service.act({
	role: 'foo',
	cmd: 'bar',
	name: 'tobi',
}, function(err, res) {
	console.log(res, Date.now());
});
```

- start `service.js`
- start `client.js`

  **client logs:** { test: 123, name: 'tobi' } 1453365970625
  **service logs:** tobi 1453365970622 bl7z8vtfzapr/pu6z54yliflt
- stop `service.js`
- start `service.js`

  **client logs:** { test: 123, name: 'tobi' } 1453366057821
  **service logs:** tobi 1453366057818 bl7z8vtfzapr/pu6z54yliflt

https://github.com/platdesign/seneca-issue

**Comment by jbruni** [comment]
I suggest calling the `close` method when the service process receives the signal to stop:
http://senecajs.org/api/#close-done-

Something like this, at **service.js**:
```
// not sure if it should be SIGINT or SIGTERM or both

process.on('SIGINT', function() {
  service.close();
});
```

**Comment by platdesign** [comment]
does not change the behaviour.

**Comment by mcdonnelldean** [comment]
Could be a bug, I'll take a look at the test case tonight if I can. @geek If you get there before me feel free to look.

**Comment by StarpTech** [comment]
@mcdonnelldean  I also have the same issue. I can reproduce it with the test scenario by @platdesign

**Comment by StarpTech** [comment]
How is the status?

**Comment by geek** [comment]
@StarpTech and @platdesign how are you restarting?

**Comment by StarpTech** [comment]
@geek here is a demonstration of the issue.
![issue](https://cloud.githubusercontent.com/assets/1764424/12872543/19f4619e-cda6-11e5-811d-2ec19bed6760.gif)

**Comment by geek** [comment]
@StarpTech thanks, thats a perfect demonstration... very clear.  I will try and get time to take a look at this over the next couple of days.

**Comment by jbruni** [comment]
At [seneca-transport readme](https://github.com/senecajs/seneca-transport#using-the-tcp-channel) we can read, about TCP channel:

> The client side of the TCP transport will also attempt to reconnect if the connection breaks, providing fault tolerance for server restarts.

Indeed, the core of the whole issue seems to be at [seneca-transport/lib/tcp.js](https://github.com/senecajs/seneca-transport/blob/master/lib/tcp.js#L96-L151), using the [reconnect-core](https://github.com/juliangruber/reconnect-core) package for the feature mentioned above.

It sends the whole set of messages again upon reconnection. This is clearly a bug. It should have a mechanism to clear messages from the queue once the original TCP request is handled.

**Comment by jbruni** [comment]
Hmmm... I've spent some more time investigating the issue...
- using lots of "**setTimeout**" and "**setInterval**" to trigger more "**service.act**" calls at the client side
- stopping/starting the service side to check what happened
- introducing several "**console.log**" calls to have a better understanding of what happens at:
  - `seneca-transport/lib/tcp.js`
  - `seneca-transport/lib/transport-utils.js`
  - `seneca-transport/transport.js` .

The "*mechanism to clear messages from the queue once the original TCP request is handled*" is somehow in place, and working... but the fact is that it only starts working after some internal "start up" job.

I've done several tests. Only the first sent message (through service.act call), or the first message plus the ones done in the next miliseconds, are re-processed upon reconnection. The messages following a given moment in time onwards are not re-processed.

Does it make sense? How can I explain it better?

Anyway: there is an **asynchronous setup process** which happens - and after it happens, everything goes as expected. **After** this setup, an eventual client reconnection does **not** trigger a re-flow of previous sent messages - as should happen all the time.

Only the first "**service.act**" call, and also the ones done in the following miliseconds, are handled **before** something from this initial setup has been settled down.

The solution for this issue involves:
- completing the full setup at "**service.client**" call moment (do not lazily wait for the first "service.act" call to perform this initialization)
- not only the above, but also somehow not allowing "service.act" calls to be processed before this initialization has been completed (they shall wait for its conclusion, and only then be processed)

**Comment by StarpTech** [comment]
@geek Could you check it?

**Comment by StarpTech** [comment]
Hi @geek it's almost one month ago since the issue has been posted. Is it taken seriously? Thanks.

**Comment by StarpTech** [comment]
@jbruni thanks for your investigation.

**Comment by geek** [comment]
@StarpTech See https://github.com/geek/seneca/blob/master/test/transport.test.js#L915-L948

Can you create a test that reproduces the issue... this test I created demonstrates that there isn't an issue when closing one server, starting another while the client runs.  

You are also welcome to open a PR to fix the issue as you see it.

**Comment by StarpTech** [comment]
@geek this test cannot cover the scenario in this test case https://github.com/platdesign/seneca-issue.
- You don't have seperate processes.
- You don't kill the process by 'CTRL+C'.
I will try to write a test but the issue you can see in my gif and you can reproduce it with the test repo.

**Comment by platdesign** [comment]
What's the status on this?

**Comment by timoweiss** [comment]
We are also experiencing this bug.
We have about 6 services (running in separate processes) and if we deploy/restart a consumer without it's producer, we get retransmitted messages.

**Comment by mcdonnelldean** [comment]
@geek Just to chime in, I don't get this issue when transporting over mesh which is why it may not be visible on nodezoo, concorda, or vidi.

**Comment by geek** [comment]
I am open to any ideas on a solution.  Right now we listen for any process `SIGHUP`, `SIGTERM`, `SIGINT`, and `SIGBREAK` signals and call `close` on the seneca instance.  This removes all listeners and executes the pattern `role: seneca, cmd: close`.

**Comment by platdesign** [comment]
@rjrodger I've looked through the implementation of seneca-transport/tcp. It seems, that `reconnect` inside of `send` is responsible for this bug. When the connection is reestablished (service.js comes up again), the callback of reconnect is called again. Wouldn't it be better if the connection is initialized outside of `send`?

@geek Its definitely a problem with the client implementation, not the listener.

**Comment by geek** [comment]
@platdesign thanks for the investigation details... I am working on a test case and resolution.

**Comment by geek** [comment]
@platdesign see https://github.com/geek/seneca-transport/commit/4433348f2df0d7b939d6580e6ed88543348fd696 for how I am reworking things.

**Comment by StarpTech** [comment]
@geek great, is it finished? When will you merge the PR ?

**Comment by geek** [comment]
@StarpTech please help review the fix: https://github.com/senecajs/seneca-transport/pull/86

**Comment by StarpTech** [comment]
@geek here are my recommendations:

- we could respect this issue hapijs/hapi#3014 to improve the performance of JSON.stringify
- we could respect circular references in JSON (instead to throw an exception) to improve availibility.

Best solution: Use http://msgpack.org/ to stream JSON over network. It is very easy to connect.

everything else is fine thank you.

**Comment by StarpTech** [comment]
@geek would be great if you can finish this issue with your PR. This issue is almost 3 month old and its a huge bug for all people who trust in the TCP transport. Any critical application run into big problems. We had to change all seneca instances to HTTP. Thanks!

**Comment by naomifeehanmoran** [comment]
@StarpTech @platdesign @timoweiss thanks for the feedback. We will get someone on the related PR to close it out asap. We understand that this issue has been open for some time and are endeavouring to resolve it as quickly as we can, so please bare with us!

**Comment by mcdonnelldean** [comment]
Reviewing this today.

**Comment by StarpTech** [comment]
@mcdonnelldean any progress?

**Comment by mcdonnelldean** [comment]
@StarpTech Yes, I've been adding progress to the PR. It looks like the new implementation works but it has a regression that stops .client() calls from being idempotent. All other tests pass so once this is done it should be ready.

**Comment by StarpTech** [comment]
@mcdonnelldean what do you think about my recommendations?

**Comment by mcdonnelldean** [comment]
I haven't considered them yet as it's not working. We are discussing transport generally in #399 Msgpack is not something that would be landed in this PR.

**Comment by mcdonnelldean** [comment]
Moved to https://github.com/senecajs/seneca-transport/issues/91

**Comment by StarpTech** [comment]
Hi @mcdonnelldean we can reproduce the issue. This is not fixed. In our production environment it leads to a big issue. Please investigate in this to make seneca basic transport stable.

Please checkout https://github.com/platdesign/seneca-issue and upgrade to the latest seneca version in the package.json e.g 2.1.0

1. Start client.json
2. Start server.json
3. Restart server.json = Server repeats the client act.

the same instruction above. Thank you.

**Comment by StarpTech** [comment]
@mcdonnelldean  Please reopen!

**Comment by mcdonnelldean** [comment]
Hey @StarpTech really sorry to hear this is still a problem. We also found the test based off the above code was returning false positives. At this stage we are looking to simply rewrite that whole section of logic. I'll ping this issue when I have more details on what the plan is.

**Comment by StarpTech** [comment]
Hi @mcdonnelldean hope you get it. Thank you.

**Comment by StarpTech** [comment]
HI @mcdonnelldean it is worth to keep at senecajs as a microservice tookit? This issue is almost 5 months old and nobody take care of it. It's not just a bad implementation it's a huge issue in the transport protocol. The fact that this simple scenario never happens on any other environemnt leads me to the conclusion that nobody use the TCP transport or senecajs isnt used anymore. 

Also the roadmap isnt updated. I thought I could expect some improvements. Plesae dont understand this wrong. Senecajs is great and you do a great job but it will become more difficult to see in senecajs a professional and robust software.

```
§ 3.0.0 - June 2016

Remove error handling in favor of bubble up methods.
Fix inconsistencies in err flow through callbacks.
Add new logging system to support object based logging.
Announce deprecation of ‘old’ logging system.
```

**Comment by mcdonnelldean** [comment]
@StarpTech

First off, apologies, I understand it can feel like nobody's listening. I want to assure you we are. Seneca is a very large project with a lot of moving parts and with that comes prioritisation.

Right now we are working on changing the roadmap. We hit some tricky core issues when we attempted to remove error handling which forced us to stop and reconsider our next steps. Logging is almost complete but we found a regression in Entities that needs to be resolved too.

Unfortunately, based on usage patterns, TCP is indeed a lesser used transport. That is not to say it should be bugged, but as with all projects resourcing is applied where most appropriate and with that comes prioritisation. Seneca is very much used, and by a large number of users who also have issues we are attending to.

One thing I would like to point out. Seneca is open source and a community project. If you feel an issue is of the utmost priority we would welcome contribution. If you want to take a swing at replacing TCP's logic with something more robust I can certainly review and help you with any issues that arise.
